modules/interactive_music/audio_stream_playlist.cpp

```diff
@@ -0,0 +1,406 @@
+	AudioServer::get_singleton()->lock();
+	stream_count = p_count;
+	AudioServer::get_singleton()->unlock();
+	notify_property_list_changed();
+}
+
+int AudioStreamPlaylist::get_stream_count() const {
+	return stream_count;
+}
+
+void AudioStreamPlaylist::set_fade_time(float p_time) {
+	fade_time = p_time;
+}
+
+float AudioStreamPlaylist::get_fade_time() const {
+	return fade_time;
+}
+
+void AudioStreamPlaylist::set_shuffle(bool p_shuffle) {
+	shuffle = p_shuffle;
+}
+
+bool AudioStreamPlaylist::get_shuffle() const {
+	return shuffle;
+}
+
+void AudioStreamPlaylist::set_loop(bool p_loop) {
+	loop = p_loop;
+}
+
+bool AudioStreamPlaylist::has_loop() const {
+	return loop;
+}
+
+void AudioStreamPlaylist::_validate_property(PropertyInfo &property) const {
+	String prop = property.name;
+	if (prop != "stream_count" && prop.begins_with("stream_")) {
+		int stream = prop.get_slicec('/', 0).get_slicec('_', 1).to_int();
+		if (stream >= stream_count) {
+			property.usage = PROPERTY_USAGE_INTERNAL;
+		}
+	}
+}
+
+void AudioStreamPlaylist::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("set_stream_count", "stream_count"), &AudioStreamPlaylist::set_stream_count);
+	ClassDB::bind_method(D_METHOD("get_stream_count"), &AudioStreamPlaylist::get_stream_count);
+
+	ClassDB::bind_method(D_METHOD("get_bpm"), &AudioStreamPlaylist::get_bpm);
+
+	ClassDB::bind_method(D_METHOD("set_list_stream", "stream_index", "audio_stream"), &AudioStreamPlaylist::set_list_stream);
+	ClassDB::bind_method(D_METHOD("get_list_stream", "stream_index"), &AudioStreamPlaylist::get_list_stream);
+
+	ClassDB::bind_method(D_METHOD("set_shuffle", "shuffle"), &AudioStreamPlaylist::set_shuffle);
+	ClassDB::bind_method(D_METHOD("get_shuffle"), &AudioStreamPlaylist::get_shuffle);
+
+	ClassDB::bind_method(D_METHOD("set_fade_time", "dec"), &AudioStreamPlaylist::set_fade_time);
+	ClassDB::bind_method(D_METHOD("get_fade_time"), &AudioStreamPlaylist::get_fade_time);
+
+	ClassDB::bind_method(D_METHOD("set_loop", "loop"), &AudioStreamPlaylist::set_loop);
+	ClassDB::bind_method(D_METHOD("has_loop"), &AudioStreamPlaylist::has_loop);
+
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "shuffle"), "set_shuffle", "get_shuffle");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "loop"), "set_loop", "has_loop");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "fade_time", PROPERTY_HINT_RANGE, "0,1,0.01,suffix:s"), "set_fade_time", "get_fade_time");
+
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "stream_count", PROPERTY_HINT_RANGE, "0," + itos(MAX_STREAMS), PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_ARRAY, "Streams,stream_,unfoldable,page_size=999,add_button_text=" + String(RTR("Add Stream"))), "set_stream_count", "get_stream_count");
+
+	for (int i = 0; i < MAX_STREAMS; i++) {
+		ADD_PROPERTYI(PropertyInfo(Variant::OBJECT, "stream_" + itos(i), PROPERTY_HINT_RESOURCE_TYPE, "AudioStream", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_INTERNAL), "set_list_stream", "get_list_stream", i);
+	}
```

Comment:
That's rather unconventional way to define properties. Why is it like this instead of using typical Vector + `get_property_list()`?

Reply:
For the most part because its easier for undo/redo, but also because it makes it safer for the running audio state if something changes in this case.

Reply:
To clarify, its best to have the least locking possible in the audio thread, so the more you can get by without locking and relying that the memory is just there even if the value has changed and its not entirely correct, the better.